src/ucp/amo/amo.inl

```diff
@@ -43,80 +88,73 @@ ucs_status_t ucp_amo_check_send_status(ucp_request_t *req, ucs_status_t status);
         \
         req->send.lane = rkey->cache.amo_lane; \
         status = _function(ep->uct_eps[req->send.lane], \
-                           UCS_PP_TUPLE_BREAK _params, \
+                           value, *result, \
                            remote_addr, rkey->cache.amo_rkey, result, \
                            &req->send.state.uct_comp); \
         return ucp_amo_check_send_status(req, status); \
     }
 
-UCP_PROGRESS_AMO_DECL(uint32_t, uct_ep_atomic_swap32, UCP_AMO_ONE_PARAM)
-
-UCP_PROGRESS_AMO_DECL(uint32_t, uct_ep_atomic_fadd32, UCP_AMO_ONE_PARAM)
-
-UCP_PROGRESS_AMO_DECL(uint32_t, uct_ep_atomic_cswap32, UCP_AMO_TWO_PARAM)
-
-UCP_PROGRESS_AMO_DECL(uint64_t, uct_ep_atomic_swap64, UCP_AMO_ONE_PARAM)
-
-UCP_PROGRESS_AMO_DECL(uint64_t, uct_ep_atomic_fadd64, UCP_AMO_ONE_PARAM)
-
-UCP_PROGRESS_AMO_DECL(uint64_t, uct_ep_atomic_cswap64, UCP_AMO_TWO_PARAM)
+#define UCP_PROGRESS_AMO_FETCH_DECL(_size)                                        \
+    static ucs_status_t _UCP_PROGRESS_AMO_FETCH(_size)(uct_pending_req_t *_self)  \
+    {                                                                             \
+        typedef uint##_size##_t type;                                             \
+                                                                                  \
+        ucp_request_t *req    = ucs_container_of(_self, ucp_request_t, send.uct); \
+        ucp_rkey_h rkey       = req->send.amo.rkey;                               \
+        ucp_ep_t *ep          = req->send.ep;                                     \
+        type value            = (type)req->send.amo.value;                        \
+        type *result          = (type *)req->send.amo.result;                     \
+        uint64_t remote_addr  = req->send.amo.remote_addr;                        \
+        uct_atomic_op_t op    = req->send.amo.uct_op;                             \
+        ucs_status_t status;                                                      \
+                                                                                  \
+        status = UCP_RKEY_RESOLVE(rkey, ep, amo);                                 \
+        if (status != UCS_OK) {                                                   \
+            return UCS_ERR_UNREACHABLE;                                           \
+        }                                                                         \
+                                                                                  \
+        req->send.lane = rkey->cache.amo_lane;                                    \
+        status = uct_ep_atomic##_size##_fetch(ep->uct_eps[req->send.lane],        \
```

Comment:
instead of a macro, let's add 'if' to check the argument size. it's not fast anyway because it's pending queue

Reply:
implemented as separate call